### DISCUSSION TITLE
What are the handles that should be exposed when writing bindings to imagemagick in another language?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5956] [Development] [answered]
### ImageMagick version

7

### Operating system, version and so on

windows/android/linux/mac/ios

### Description

Hello,

As I've mentioned several times before I am creating bindings for the MagickWand API in Dart. 

My approach is simple, I am writing a mapping for each function of [MagickWand API](https://imagemagick.org/script/magick-wand.php) by creating an equivalent function in Dart. And I have mapped quite alot of functions already.

So I had to create equivalent structures to the C `struct`s used by the MagickWand API, specifically; `MagickWand`, `DrawingWand`, and `PixelWand`.

I just have some small questions, since I am not very familliar with the usage **of every function** in the API:

1 - do you recommend other `struct`s to be exposed to the user of the API, like the `Image` `struct` for example?
 I would like to know what would you recommend to have an idea about the things I should implement other than the functions on the [magick wand reference page](https://imagemagick.org/script/magick-wand.php) and the three structs mentioned above.
2 - Are the `Image` struct and other similar `struct`s intended to be manipulated directly by the user of the MagickWand API, or are they intended to be manipulated only by the MagickWand API methods? (I know the user in C can do whatever he wants, but I am asking about the recommended way of doing things in MagickWand API)
For example, for a user of the MagickWand (and not MagickCore), should he be creating an `Image` and setting its attributes like 

```
Image* image = malloc(sizeof *image);
image->compression = someCompressionType;
```
Or should he do:
```
MagickWand* wand = NewMagickWand();
MagickReadImage(wand, "some image");
MagickSetCompression(someCompression);
```
?

Thank you.

**Comment by urban-warrior** [comment]
MagickWand was designed to hide the complexities of the MagickCore low-level interface, as such, it uses setters/getters negating the need for direct access to the underlying MagickCore structures such as `Image` (per your example).  Just bind MagickWand and if someone has a need to get at an `Image` member that does not currently have an equivalent setter/getter, the user should instead request a mod to the MagickWand API which can then be reflected in the Dart bindings.

**Reply by Haidar0096** [reply]
@urban-warrior thanks for clarfication.

So as a small follow up question, can I conclude that methods such as
```
Image *GetImageFromMagickWand(const MagickWand *)
```

which are present in the MagickWand API, are better to NOT be mapped since they expose handles that are not recommended to be manipulated manually?

**Comment by urban-warrior** [comment, marked as answer]
Those methods are a bridge between MagickWand and MagickCore.  If you want to create that bridge, support GetImageFromMagickWand()  -- but that would mean you would need to create bindings between Dart and MagickCore as well, which would be a daunting task.

**Reply by Haidar0096** [reply]
@urban-warrior 100%, I was looking at the `Image` struct and it doesn't look easy to keep track of all those fields and their pointers, so I will stick to mapping/supporting only MagickWand methods.

 Now with this knowledge the rest of the bindings are straightforward, and I will let you know when they are done if you are interested to link them in the docs.

Again thanks for your patience :)

**Reply by dlemstra** [reply]
You might also consider using this library that I created: https://github.com/dlemstra/magick-wasm. It is a webassembly library that use the ImageMagick library.
